rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -241,11 +252,16 @@ class PlayerImpl
     std::shared_ptr<rclcpp::GenericPublisher> publisher_;
     std::function<void(const rclcpp::SerializedMessage &)> publish_func_;
   };
+
   bool is_ready_to_play_from_queue_{false};
   std::mutex ready_to_play_from_queue_mutex_;
   std::condition_variable ready_to_play_from_queue_cv_;
   rclcpp::Publisher<rosgraph_msgs::msg::Clock>::SharedPtr clock_publisher_;
-  std::unordered_map<std::string, std::shared_ptr<PlayerImpl::PlayerPublisher>> publishers_;
+  using SharedPlayerPublisher = std::shared_ptr<PlayerPublisher>;
+  using SharedPlayerClient = std::shared_ptr<PlayerServiceClient>;
+  std::unordered_map<
+    std::string,
+    std::variant<SharedPlayerPublisher, SharedPlayerClient>> senders_;
```

Comment:
```suggestion
  using PlayerPublisherSharedPtr = std::shared_ptr<PlayerPublisher>;
  using PlayerServiceClientSharedPtr = std::shared_ptr<PlayerServiceClient>;
  std::unordered_map<std::string, PlayerPublisherSharedPtr> publishers_
  std::unordered_map<std::string, PlayerServiceClientSharedPtr>> services_clients_;
```
Prefixing with `Shared` may be misleading as the entity using from multiple threads or objects, but in fact, it is just a `shared_pointer`. We used to use the `SharedPtr` postfix instead for shared_pointer types definitions.

@Barry-Xu-2018 Please, please don't use `variant` for mixing `ServicesClients` with `Publishers`. This are two different entities with completely different processing logic. Using variant in this particular place is overengineering and significantly overcomplicating the code and logic. We will have to have 100500 extra checks to make sure that we selecting the right type. The logic will be much simpler if keep `ServicesClients` and `Publishers` in separate maps.

Reply:
Thanks, updated

Reply:
@Barry-Xu-2018 I see that you mentioned that you addressed my comments but I don't really see that they were been addressed on the latest version of your branch. Did you miss a commit or overwrite it with something?

Reply:
@MichaelOrlov 

Related change has been submitted.
Please look at https://github.com/Barry-Xu-2018/rosbag2/blob/review/rosbag2_service_play/rosbag2_transport/src/rosbag2_transport/player.cpp#L266-L269

Changed commit ID is [56fffd9](https://github.com/ros2/rosbag2/pull/1481/commits/56fffd9cdb5602f930efa12fda0f7692efff0924)

Reply:
@Barry-Xu-2018 Thanks you are right. It seems I messed up wit revisions.